relay/cmd/main.go

```diff
@@ -0,0 +1,196 @@
+	rootCmd.PersistentFlags().StringVarP(&cobraConfig.TlsKeyFile, "tls-key-file", "k", "", "")
+	rootCmd.PersistentFlags().StringVarP(&cobraConfig.AuthSecret, "auth-secret", "s", "", "log level")
+}
+
+func waitForExitSignal() {
+	osSigs := make(chan os.Signal, 1)
+	signal.Notify(osSigs, syscall.SIGINT, syscall.SIGTERM)
+	<-osSigs
+}
+
+func loadConfig(configFile string) (*Config, error) {
+	log.Infof("loading config from: %s", configFile)
+	loadedConfig := &Config{}
+	_, err := util.ReadJson(configFile, loadedConfig)
+	if err != nil {
+		return nil, err
+	}
+	if cobraConfig.ListenAddress != "" {
+		loadedConfig.ListenAddress = cobraConfig.ListenAddress
+	}
+
+	if cobraConfig.ExposedAddress != "" {
+		loadedConfig.ExposedAddress = cobraConfig.ExposedAddress
+	}
+	if cobraConfig.LetsencryptDataDir != "" {
+		loadedConfig.LetsencryptDataDir = cobraConfig.LetsencryptDataDir
+	}
+	if len(cobraConfig.LetsencryptDomains) > 0 {
+		loadedConfig.LetsencryptDomains = cobraConfig.LetsencryptDomains
+	}
+	if cobraConfig.TlsCertFile != "" {
+		loadedConfig.TlsCertFile = cobraConfig.TlsCertFile
+	}
+	if cobraConfig.TlsKeyFile != "" {
+		loadedConfig.TlsKeyFile = cobraConfig.TlsKeyFile
+	}
+	if cobraConfig.AuthSecret != "" {
+		loadedConfig.AuthSecret = cobraConfig.AuthSecret
+	}
+
+	return loadedConfig, err
+}
+
+func execute(cmd *cobra.Command, args []string) error {
+	cfg, err := loadConfig(cfgFile)
+	if err != nil {
+		return fmt.Errorf("failed to load config: %s", err)
+	}
+
+	err = cfg.Validate()
+	if err != nil {
+		log.Errorf("invalid config: %s", err)
+		os.Exit(1)
+	}
+
+	metricsServer, err := metrics.NewServer(metricsPort, "")
+	if err != nil {
+		return fmt.Errorf("setup metrics: %v", err)
+	}
+
+	go func() {
+		log.Infof("running metrics server: %s%s", metricsServer.Addr, metricsServer.Endpoint)
+		if err := metricsServer.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
+			log.Fatalf("Failed to start metrics server: %v", err)
+		}
+	}()
+
+	srvListenerCfg := server.ListenerConfig{
+		Address: cfg.ListenAddress,
+	}
+	if cfg.HasLetsEncrypt() {
+		tlsCfg, err := setupTLSCertManager(cfg.LetsencryptDataDir, cfg.LetsencryptDomains...)
+		if err != nil {
+			return fmt.Errorf("%s", err)
+		}
+		srvListenerCfg.TLSConfig = tlsCfg
+	} else if cfg.HasCertConfig() {
+		tlsCfg, err := encryption.LoadTLSConfig(cfg.TlsCertFile, cfg.TlsKeyFile)
+		if err != nil {
+			return fmt.Errorf("%s", err)
+		}
+		srvListenerCfg.TLSConfig = tlsCfg
+	}
+
+	tlsSupport := srvListenerCfg.TLSConfig != nil
+
+	authenticator := auth.NewTimedHMACValidator(cfg.AuthSecret, 24*time.Hour)
+	srv, err := server.NewServer(metricsServer.Meter, cfg.ExposedAddress, tlsSupport, authenticator)
+	if err != nil {
+		return fmt.Errorf("failed to create relay server: %v", err)
+	}
+	log.Infof("server will be available on: %s", srv.InstanceURL())
+	err = srv.Listen(srvListenerCfg)
+	if err != nil {
+		return fmt.Errorf("failed to bind server: %s", err)
+	}
+
+	waitForExitSignal()
```

Comment:
This here is ineffective because it only starts intercepting signals after srv.Listen has already returned.

Something like this would work

```
	osSigs := make(chan os.Signal, 1)
	signal.Notify(osSigs, syscall.SIGINT, syscall.SIGTERM)
	
	go func() {
		err = srv.Listen(srvListenerCfg)
		if err != nil {
			log.Fatalf("failed to start server: %v", err)
		}
	}()

	<-osSigs
	err = srv.Close()
	if err != nil {
		return fmt.Errorf("failed to close server: %s", err)
	}
```

Or move srv.Close in a goroutine and leave srv.Listen in the main goroutine.
(Plus improve reporting errors)